python/lsst/ts/love/csc/authorization_model.py

```diff
@@ -0,0 +1,180 @@
+            "non_authorized_cscs"
+        ].union(non_authorized_cscs_to_add)
+        component_authorization["non_authorized_cscs"] -= non_authorized_cscs_to_remove
+
+        return component_authorization
+
+    def set_authorization(self, component, authorized_users, non_authorized_cscs):
+        """Set authorization for a component with the input values.
+
+        Parameters
+        ----------
+        component : `str`
+            Name of the component to set authorization.
+        authorized_users : `list` of `str`
+            List of strings with the name of the users to authorize.
+        non_authorized_cscs : `list` of `str`
+            List of strings with the name of CSCs to remove authorization.
+        """
+        self._handle_enter_component_in_auth_list(component=component)
+
+        self.authorization[component]["authorized_users"] = set(authorized_users)
+        self.authorization[component]["non_authorized_cscs"] = set(non_authorized_cscs)
+
+        self._handle_exit_component_in_auth_list(component=component)
+
+    def set_authorization_mode(self, auto):
+        """Set authorization mode.
+
+        Parameters
+        ----------
+        auto : `bool`
+            Set automatic mode?
+        """
+
+        self._auto_accept_request = auto
+
+    def get_component_authorization(self, component):
+        return self.authorization.get(
+            component, dict(authorized_users=set(), non_authorized_cscs=set())
+        )
+
+    async def process_request(self, data):
+        """Process request to change authorization.
+
+        Parameters
+        ----------
+        data : `object`
+            DDS-like object with the request to change authorization.
+
+        Yields
+        ------
+        set_auth_list : `dict`
+            Dictionary with the name of the component and payload for the new
+            auth list.
+        """
+
+        if not self._auto_accept_request:
+            raise NotImplementedError("Non-automatic mode not implemented yet.")
+
+        for component in data.cscsToChange.split(","):
+
+            component_authorizaton = self.get_authorization(
+                component=component,
+                authorized_users=data.authorizedUsers,
+                non_authorized_cscs=data.nonAuthorizedCSCs,
+            )
+
+            self.log.debug(f"Component authorization: {component_authorizaton}")
+
+            set_auth_list = dict(
+                component=component,
+                auth_list=dict(
+                    authorizedUsers=",".join(
+                        component_authorizaton["authorized_users"]
+                    ),
+                    nonAuthorizedCSCs=",".join(
+                        component_authorizaton["non_authorized_cscs"]
+                    ),
+                ),
+            )
+
+            yield set_auth_list
+
+    def _handle_enter_component_in_auth_list(self, component):
```

Comment:
Please consider a context manager for this and the associated exit method. It seems tailor-made for this purpose: you want to perform an action when starting a code block and perform another action when the code block is done.

This makes the API easier to use correctly (the user doesn't have to remember to call exit) and it gives you an opportunity to handle exceptions in the code block cleanly, should you want that.

contextlib.contextmanager makes these trivial to write. You do your initial work in a method, yield, then do your cleanup work. The user just says:
```
with self.authlist_context(component):
    do stuff
```
The result should be fewer lines of code than you have now, and be easier to read.